{"cells":[{"cell_type":"code","source":["from google.colab import drive\n","drive.mount('/content/drive')"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"kBLbcd-9cZmE","executionInfo":{"status":"ok","timestamp":1688802439870,"user_tz":420,"elapsed":2394,"user":{"displayName":"Jiayi Ni","userId":"15636163276706872048"}},"outputId":"a6f29220-f868-4343-cd06-0e0767c909bd"},"id":"kBLbcd-9cZmE","execution_count":9,"outputs":[{"output_type":"stream","name":"stdout","text":["Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n"]}]},{"cell_type":"code","execution_count":10,"id":"34cad001","metadata":{"id":"34cad001","executionInfo":{"status":"ok","timestamp":1688802442728,"user_tz":420,"elapsed":296,"user":{"displayName":"Jiayi Ni","userId":"15636163276706872048"}}},"outputs":[],"source":["import pandas as pd\n","import numpy as np\n","import scipy\n","from sklearn.preprocessing import StandardScaler\n","from bed_reader import open_bed, sample_file"]},{"cell_type":"code","execution_count":11,"id":"220d3bc3","metadata":{"id":"220d3bc3","executionInfo":{"status":"ok","timestamp":1688802443938,"user_tz":420,"elapsed":4,"user":{"displayName":"Jiayi Ni","userId":"15636163276706872048"}}},"outputs":[],"source":["def simulator(X, sigma):\n","    '''\n","    Simulate phenotype y from X and sigmas\n","    '''\n","    N = X.shape[0]\n","    M = X.shape[1]\n","    y = np.zeros((N,1))\n","    sigma_epsilon=1 - sigma # environmental effect sizes\n","    betas = np.random.randn(M,1)*np.sqrt(sigma) # additive SNP effect sizes\n","    y += X@betas/np.sqrt(M)\n","    #print(f'sigma_epislon={sigma_epsilon}')\n","    y += np.random.randn(N,1)*np.sqrt(sigma_epsilon) # add the effect sizes\n","    return y, betas\n","\n","def solve_linear_equation(X, y):\n","    '''\n","    Solve least square\n","    '''\n","    sigma = np.linalg.lstsq(X, y, rcond=None)[0]\n","    return sigma\n","\n","\n","def solve_linear_qr(X, y):\n","    '''\n","    Solve least square using QR decomposition\n","    '''\n","    Q, R = scipy.linalg.qr(X)\n","    sigma = scipy.linalg.solve_triangular(R, np.dot(Q.T, y))\n","    return sigma\n","\n","def RHE(X,y,num_random_vect=10,seed=1,verbose=False):\n","    '''\n","    RHE estimation\n","    '''\n","    np.random.seed(seed)\n","    N = X.shape[0]\n","    M = X.shape[1]\n","    T = np.zeros((2,2))\n","    q = np.zeros((2,1))\n","\n","    Xi = X.copy()/np.sqrt(M)\n","    for _ in range(num_random_vect):\n","        # Generate random vector to estimate trace\n","        rand_vector = np.random.randn(N,1)\n","        T[0,0] += rand_vector.T@Xi@Xi.T@Xi@Xi.T@rand_vector/num_random_vect\n","\n","    T[1,0] = np.trace(Xi@Xi.T)\n","    T[0,1] = T[1,0]\n","    T[1,1] = N\n","\n","    q[0] = y.T@Xi@Xi.T@y\n","    q[1] = y.T@y\n","\n","    if verbose:\n","        print(T)\n","    sigma_est = solve_linear_equation(T,q)\n","    return sigma_est\n","\n","\n","def HE (X, y, verbose = False):\n","    '''\n","    HE estimation (without using random vectors)\n","    '''\n","    N = X.shape[0]\n","    M = X.shape[1]\n","    T = np.zeros((2,2))\n","    q = np.zeros((2,1))\n","\n","    Xi = X.copy()/np.sqrt(M)\n","    T[0,0] = np.trace(Xi@Xi.T@Xi@Xi.T)\n","    T[1,0] = np.trace(Xi@Xi.T)\n","    T[0,1] = T[1,0]\n","    T[1,1] = N\n","\n","    q[0] = y.T@Xi@Xi.T@y\n","    q[1] = y.T@y\n","\n","    if verbose:\n","        print(T)\n","\n","    sigma_est = solve_linear_equation(T,q)\n","    return sigma_est"]},{"cell_type":"markdown","id":"f60f735f","metadata":{"id":"f60f735f"},"source":["## In the cell below, we simulate the genotype matrix by taking random samples of the MAF for each SNP from a uniform distribution. Then for each individual (based on the MAF of that SNP), run a Bernoulli trial to get the genotype value at that location (i.e., whether the individual has the SNP or not; either 0,1 or 2)\n","I'll also share the genotype matrices for you to benchmark RHE and the python version (here), but this a very simple simulation. For instance, here we don't consider the LD structure here at all."]},{"cell_type":"code","execution_count":12,"id":"f794da77","metadata":{"id":"f794da77","outputId":"e62c99d6-fef2-4536-af12-d371025a0c11","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1688802451726,"user_tz":420,"elapsed":5392,"user":{"displayName":"Jiayi Ni","userId":"15636163276706872048"}}},"outputs":[{"output_type":"stream","name":"stdout","text":["Actual sigmas are: 0.3\n","RHE estimated sigmas are: [0.27646048] [0.71014217] \n","HE estimated sigmas are: [0.28039846] [0.7062042]\n"]}],"source":["M=100 # M = 500000\n","N=10000  # N=300000\n","target_index=1\n","sigmas_list=[0.3]\n","# t0 = time()\n","for i in range(1):\n","    np.random.seed(i)\n","    sigmas=sigmas_list[i]\n","\n","    MAFs = np.random.uniform(0.01,0.5,M)\n","\n","    P_matrix = np.tile(MAFs, (N, 1))\n","\n","    X = np.random.binomial(2, P_matrix) # X here is the un-standardized, raw genotype matrix (entries are 0 or 1)\n","\n","    X = StandardScaler().fit_transform(X) # standardize X\n","\n","    y, beta_list = simulator(X, sigmas) # generate phenotype with some beta values\n","\n","    print(f'Actual sigmas are: {sigmas}')\n","    sigma_est=RHE(X,y,num_random_vect=10,seed=42) # run py-RHE\n","    sigma_est2=HE(X, y)\n","\n","    print ('RHE estimated sigmas are:',*sigma_est, '\\nHE estimated sigmas are:', *sigma_est2)"]},{"cell_type":"markdown","id":"703ba44d","metadata":{"id":"703ba44d"},"source":["## And here's an (almost) identical code but instead of simulating genotype, it reads in the genotype file (.bed) using the bed-reader. One major difference is, in real genotype, there is often missingness, which needs to be imputed.\n","### Here's one way of imputing: simply replace the missing values by the average of non-missing entries for that SNP. This is different than what's done in the RHE code.\n"]},{"cell_type":"code","execution_count":13,"id":"e540a2a2","metadata":{"id":"e540a2a2","executionInfo":{"status":"ok","timestamp":1688802457475,"user_tz":420,"elapsed":295,"user":{"displayName":"Jiayi Ni","userId":"15636163276706872048"}}},"outputs":[],"source":["'''\n","impute the genotype matrix for any missingness by using the average of other samples\n","this function also standardizes it\n","'''\n","def impute_geno(X):\n","    N = X.shape[0]\n","    M = X.shape[1]\n","    X_imp = X.copy()\n","    for m in range(M):\n","        cnt = 0\n","        csum = 0\n","        for n in range(N):\n","            if not np.isnan(X[n, m]):\n","                cnt += 1\n","                csum += X[n,m]\n","        csum /= cnt\n","        X_imp[:,m] = np.nan_to_num(X_imp[:,m], nan=csum*0.5)\n","    X_imp = (X_imp-np.mean(X_imp, axis=0))/np.std(X_imp, axis=0)\n","    return X_imp"]},{"cell_type":"code","execution_count":15,"id":"e0a3553b","metadata":{"id":"e0a3553b","outputId":"4597c676-c348-4ec5-df9a-125e378e9d6b","colab":{"base_uri":"https://localhost:8080/","height":389},"executionInfo":{"status":"error","timestamp":1688802602637,"user_tz":420,"elapsed":307,"user":{"displayName":"Jiayi Ni","userId":"15636163276706872048"}}},"outputs":[{"output_type":"error","ename":"FileNotFoundError","evalue":"ignored","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)","\u001b[0;32m<ipython-input-15-bc1be6de8460>\u001b[0m in \u001b[0;36m<cell line: 3>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mgeno_path\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"/content/drive/MyDrive/Colab Notebooks/RHE_project/geno/1kindv_1ksnps.bed\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mbed\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen_bed\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgeno_path\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mX\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mbed\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/bed_reader/_open_bed.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(self, index, dtype, order, force_python_only, num_threads)\u001b[0m\n\u001b[1;32m    366\u001b[0m         \u001b[0;31m# allocation them?\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    367\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_iid_range\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 368\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_iid_range\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miid_count\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"intp\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    369\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sid_range\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    370\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sid_range\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msid_count\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"intp\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/bed_reader/_open_bed.py\u001b[0m in \u001b[0;36miid_count\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    925\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    926\u001b[0m         \"\"\"\n\u001b[0;32m--> 927\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_count\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"fam\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    928\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    929\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mproperty\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/bed_reader/_open_bed.py\u001b[0m in \u001b[0;36m_count\u001b[0;34m(self, suffix)\u001b[0m\n\u001b[1;32m    965\u001b[0m         \u001b[0mcount\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_counts\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0msuffix\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    966\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mcount\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 967\u001b[0;31m             \u001b[0mcount\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_rawincount\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_property_filepath\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msuffix\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    968\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_counts\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0msuffix\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcount\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    969\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mcount\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/bed_reader/_open_bed.py\u001b[0m in \u001b[0;36m_rawincount\u001b[0;34m(filepath)\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[0;31m# https://stackoverflow.com/questions/845058/how-to-get-line-count-of-a-large-file-cheaply-in-python\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0m_rawincount\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 17\u001b[0;31m     \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"rb\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     18\u001b[0m         \u001b[0mbufgen\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtakewhile\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mraw\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1024\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0;36m1024\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0m_\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrepeat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     19\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbuf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcount\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mb\"\\n\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mbuf\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mbufgen\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/content/drive/MyDrive/Colab Notebooks/RHE_project/geno/1kindv_1ksnps.fam'"]}],"source":["geno_path=\"/content/drive/MyDrive/Colab Notebooks/RHE_project/geno/1kindv_1ksnps.bed\"\n","bed = open_bed(geno_path)\n","X = bed.read()\n","print(X.shape)"]},{"cell_type":"code","execution_count":null,"id":"63e3bcce","metadata":{"id":"63e3bcce"},"outputs":[],"source":["X_imp = impute_geno(X)"]},{"cell_type":"code","execution_count":null,"id":"c5fda16b","metadata":{"id":"c5fda16b","outputId":"ed054a68-9e43-42e5-b1ac-5d283970e1e1"},"outputs":[{"name":"stdout","output_type":"stream","text":["Actual sigmas are: 0.2\n","RHE estimated sigmas are: [0.16427573] [0.831989] \n","HE estimated sigmas are: [0.16982968] [0.82643505]\n"]}],"source":["sigma = 0.2\n","y, beta_list = simulator(X_imp, sigma) # generate phenotype with some beta values\n","\n","print(f'Actual sigmas are: {sigma}')\n","sigma_est=RHE(X_imp,y,num_random_vect=10,seed=42) # run py-RHE\n","sigma_est2=HE(X_imp, y) # run exact-HE\n","\n","print ('RHE estimated sigmas are:',*sigma_est, '\\nHE estimated sigmas are:', *sigma_est2)"]},{"cell_type":"markdown","id":"41ed1dfd","metadata":{"id":"41ed1dfd"},"source":["## TODO: In the \"geno\" directory, you will find several genotype (.bed) files with multiple combinations of N (samples) and M (snps). As discussed in our meeting:\n","\n","## 1. Benchmark the python version of RHE & HE and the original, C++ version on those genotypes files and plot the run time (y-axis) against NxM (x-axis), and interpolate to large samples (UKBB) of 300k individuals x 454k SNPs.\n","\n","## 2. Repeat the same, but using the simulated genotypes\n","For simulated phenotypes, to run it on the C++ RHE, you'd have to port the matrix into PLINK bed. Read https://zzz.bwh.harvard.edu/plink/binary.shtml for how the encoding works. You might also want to take a look at this documentation for packages that does it: https://rdrr.io/cran/genio/man/write_bed.html"]},{"cell_type":"markdown","id":"65a00019","metadata":{"id":"65a00019"},"source":["## 3*.  If you are able to do these and get some sensible results, one extra task is to take a look at how RHE imputes & standardizes the genotype matrix, and see if you can implement it in python"]},{"cell_type":"code","execution_count":null,"id":"703ad33d","metadata":{"id":"703ad33d"},"outputs":[],"source":[]}],"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.9.13"},"colab":{"provenance":[]}},"nbformat":4,"nbformat_minor":5}